File: webapp/utils/validation.py
#!/usr/bin/env python3
"""
Input Validation Utilities for API Endpoints
Comprehensive validation for all API inputs
"""

from typing import Dict, Any, Optional, Tuple, List


def validate_user_id(user_id: Any) -> Tuple[bool, Optional[str]]:
    """
    Validate user ID format
    
    Args:
        user_id: User ID to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not user_id:
        return False, 'User ID is required'
    if not isinstance(user_id, str):
        return False, 'User ID must be a string'
    if len(user_id) < 3 or len(user_id) > 100:
        return False, 'User ID must be between 3 and 100 characters'
    if not user_id.replace('_', '').replace('-', '').isalnum():
        return False, 'User ID contains invalid characters'
    return True, None


def validate_session_id(session_id: Any) -> Tuple[bool, Optional[str]]:
    """
    Validate session ID format
    
    Args:
        session_id: Session ID to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not session_id:
        return False, 'Session ID is required'
    if not isinstance(session_id, str):
        return False, 'Session ID must be a string'
    if len(session_id) < 10 or len(session_id) > 200:
        return False, 'Session ID must be between 10 and 200 characters'
    return True, None


def validate_post_content(content: Any, max_length: int = 10000) -> Tuple[bool, Optional[str]]:
    """
    Validate post content
    
    Args:
        content: Post content to validate
        max_length: Maximum length
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not content:
        return False, 'Content is required'
    if not isinstance(content, str):
        return False, 'Content must be a string'
    if len(content) > max_length:
        return False, f'Content must be no more than {max_length} characters'
    if not content.strip():
        return False, 'Content cannot be empty'
    return True, None


def validate_post_id(post_id: Any) -> Tuple[bool, Optional[str]]:
    """
    Validate post ID format
    
    Args:
        post_id: Post ID to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not post_id:
        return False, 'Post ID is required'
    if not isinstance(post_id, str):
        return False, 'Post ID must be a string'
    if len(post_id) < 5 or len(post_id) > 100:
        return False, 'Post ID must be between 5 and 100 characters'
    if not post_id.replace('_', '').replace('-', '').isalnum():
        return False, 'Post ID contains invalid characters'
    return True, None


def validate_pagination(limit: Any, offset: Any, max_limit: int = 100) -> Tuple[bool, Optional[str], Optional[int], Optional[int]]:
    """
    Validate pagination parameters
    
    Args:
        limit: Items per page
        offset: Offset
        max_limit: Maximum limit
        
    Returns:
        Tuple of (is_valid, error_message, limit, offset)
    """
    try:
        limit_num = int(limit)
    except (ValueError, TypeError):
        return False, 'Limit must be an integer', None, None
    
    try:
        offset_num = int(offset)
    except (ValueError, TypeError):
        return False, 'Offset must be an integer', None, None
    
    if limit_num < 1:
        return False, 'Limit must be a positive integer', None, None
    if limit_num > max_limit:
        return False, f'Limit cannot exceed {max_limit}', None, None
    if offset_num < 0:
        return False, 'Offset must be a non-negative integer', None, None
    
    return True, None, limit_num, offset_num


def validate_feed_type(feed_type: Any) -> Tuple[bool, Optional[str]]:
    """
    Validate feed type
    
    Args:
        feed_type: Feed type to validate
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    valid_types = ['chronological', 'quality', 'personalized']
    if not feed_type:
        return False, 'Feed type is required'
    if feed_type not in valid_types:
        return False, f'Feed type must be one of: {", ".join(valid_types)}'
    return True, None


def validate_comment_content(content: Any, max_length: int = 5000) -> Tuple[bool, Optional[str]]:
    """
    Validate comment content
    
    Args:
        content: Comment content
        max_length: Maximum length
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not content:
        return False, 'Comment content is required'
    if not isinstance(content, str):
        return False, 'Comment content must be a string'
    if len(content) > max_length:
        return False, f'Comment must be no more than {max_length} characters'
    if not content.strip():
        return False, 'Comment cannot be empty'
    return True, None


def validate_media(media: Any, max_items: int = 10) -> Tuple[bool, Optional[str]]:
    """
    Validate media array
    
    Args:
        media: Media array
        max_items: Maximum items
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not isinstance(media, list):
        return False, 'Media must be an array'
    if len(media) > max_items:
        return False, f'Cannot attach more than {max_items} media items'
    
    for item in media:
        if not isinstance(item, dict):
            return False, 'Each media item must be an object'
        if 'type' not in item or 'url' not in item:
            return False, 'Each media item must have type and url'
        if item['type'] not in ['image', 'video']:
            return False, 'Media type must be "image" or "video"'
        if not isinstance(item['url'], str) or not item['url']:
            return False, 'Media URL must be a non-empty string'
    
    return True, None


def validate_tags(tags: Any, max_items: int = 20) -> Tuple[bool, Optional[str]]:
    """
    Validate tags array
    
    Args:
        tags: Tags array
        max_items: Maximum items
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not isinstance(tags, list):
        return False, 'Tags must be an array'
    if len(tags) > max_items:
        return False, f'Cannot use more than {max_items} tags'
    
    for tag in tags:
        if not isinstance(tag, str):
            return False, 'Each tag must be a string'
        if len(tag) == 0 or len(tag) > 50:
            return False, 'Each tag must be between 1 and 50 characters'
        # Allow alphanumeric, underscore, hyphen, and #
        if not all(c.isalnum() or c in ['_', '-', '#'] for c in tag):
            return False, 'Tags can only contain letters, numbers, underscores, hyphens, and #'
    
    return True, None


def validate_visibility(visibility: Any) -> Tuple[bool, Optional[str]]:
    """
    Validate visibility setting
    
    Args:
        visibility: Visibility setting
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    valid_values = ['public', 'followers', 'private']
    if not visibility:
        return False, 'Visibility is required'
    if visibility not in valid_values:
        return False, f'Visibility must be one of: {", ".join(valid_values)}'
    return True, None


def validate_message(message: Any, max_length: int = 10000) -> Tuple[bool, Optional[str]]:
    """
    Validate message content (for Thesidia API)
    
    Args:
        message: Message content
        max_length: Maximum length
        
    Returns:
        Tuple of (is_valid, error_message)
    """
    if not message:
        return False, 'Message is required'
    if not isinstance(message, str):
        return False, 'Message must be a string'
    if len(message) > max_length:
        return False, f'Message must be no more than {max_length} characters'
    if not message.strip():
        return False, 'Message cannot be empty'
    return True, None

File: webapp/utils/validation.js
/**
 * Input Validation Utilities
 * Comprehensive validation for all API inputs
 */

/**
 * Validate user ID format
 * @param {string} user_id - User ID to validate
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateUserId(user_id) {
    if (!user_id) {
        return { valid: false, error: 'User ID is required' };
    }
    if (typeof user_id !== 'string') {
        return { valid: false, error: 'User ID must be a string' };
    }
    if (user_id.length < 3 || user_id.length > 100) {
        return { valid: false, error: 'User ID must be between 3 and 100 characters' };
    }
    if (!/^[a-zA-Z0-9_-]+$/.test(user_id)) {
        return { valid: false, error: 'User ID contains invalid characters' };
    }
    return { valid: true };
}

/**
 * Validate session ID format
 * @param {string} session_id - Session ID to validate
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateSessionId(session_id) {
    if (!session_id) {
        return { valid: false, error: 'Session ID is required' };
    }
    if (typeof session_id !== 'string') {
        return { valid: false, error: 'Session ID must be a string' };
    }
    if (session_id.length < 10 || session_id.length > 200) {
        return { valid: false, error: 'Session ID must be between 10 and 200 characters' };
    }
    return { valid: true };
}

/**
 * Validate post content
 * @param {string} content - Post content to validate
 * @param {number} maxLength - Maximum length (default 10000)
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validatePostContent(content, maxLength = 10000) {
    if (!content) {
        return { valid: false, error: 'Content is required' };
    }
    if (typeof content !== 'string') {
        return { valid: false, error: 'Content must be a string' };
    }
    if (content.length > maxLength) {
        return { valid: false, error: `Content must be no more than ${maxLength} characters` };
    }
    if (content.trim().length === 0) {
        return { valid: false, error: 'Content cannot be empty' };
    }
    return { valid: true };
}

/**
 * Validate post ID format
 * @param {string} post_id - Post ID to validate
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validatePostId(post_id) {
    if (!post_id) {
        return { valid: false, error: 'Post ID is required' };
    }
    if (typeof post_id !== 'string') {
        return { valid: false, error: 'Post ID must be a string' };
    }
    if (post_id.length < 5 || post_id.length > 100) {
        return { valid: false, error: 'Post ID must be between 5 and 100 characters' };
    }
    if (!/^[a-zA-Z0-9_-]+$/.test(post_id)) {
        return { valid: false, error: 'Post ID contains invalid characters' };
    }
    return { valid: true };
}

/**
 * Validate pagination parameters
 * @param {number} limit - Items per page
 * @param {number} offset - Offset
 * @param {number} maxLimit - Maximum limit (default 100)
 * @returns {Object} {valid: boolean, error?: string, limit?: number, offset?: number}
 */
export function validatePagination(limit, offset, maxLimit = 100) {
    const limitNum = parseInt(limit, 10);
    const offsetNum = parseInt(offset, 10);
    
    if (isNaN(limitNum) || limitNum < 1) {
        return { valid: false, error: 'Limit must be a positive integer' };
    }
    if (limitNum > maxLimit) {
        return { valid: false, error: `Limit cannot exceed ${maxLimit}` };
    }
    if (isNaN(offsetNum) || offsetNum < 0) {
        return { valid: false, error: 'Offset must be a non-negative integer' };
    }
    
    return { valid: true, limit: limitNum, offset: offsetNum };
}

/**
 * Validate feed type
 * @param {string} feed_type - Feed type to validate
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateFeedType(feed_type) {
    const validTypes = ['chronological', 'quality', 'personalized'];
    if (!feed_type) {
        return { valid: false, error: 'Feed type is required' };
    }
    if (!validTypes.includes(feed_type)) {
        return { valid: false, error: `Feed type must be one of: ${validTypes.join(', ')}` };
    }
    return { valid: true };
}

/**
 * Validate comment content
 * @param {string} content - Comment content
 * @param {number} maxLength - Maximum length (default 5000)
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateCommentContent(content, maxLength = 5000) {
    if (!content) {
        return { valid: false, error: 'Comment content is required' };
    }
    if (typeof content !== 'string') {
        return { valid: false, error: 'Comment content must be a string' };
    }
    if (content.length > maxLength) {
        return { valid: false, error: `Comment must be no more than ${maxLength} characters` };
    }
    if (content.trim().length === 0) {
        return { valid: false, error: 'Comment cannot be empty' };
    }
    return { valid: true };
}

/**
 * Validate media array
 * @param {Array} media - Media array
 * @param {number} maxItems - Maximum items (default 10)
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateMedia(media, maxItems = 10) {
    if (!Array.isArray(media)) {
        return { valid: false, error: 'Media must be an array' };
    }
    if (media.length > maxItems) {
        return { valid: false, error: `Cannot attach more than ${maxItems} media items` };
    }
    
    for (const item of media) {
        if (typeof item !== 'object' || !item.type || !item.url) {
            return { valid: false, error: 'Each media item must have type and url' };
        }
        if (!['image', 'video'].includes(item.type)) {
            return { valid: false, error: 'Media type must be "image" or "video"' };
        }
        if (typeof item.url !== 'string' || item.url.length === 0) {
            return { valid: false, error: 'Media URL must be a non-empty string' };
        }
    }
    
    return { valid: true };
}

/**
 * Validate tags array
 * @param {Array} tags - Tags array
 * @param {number} maxItems - Maximum items (default 20)
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateTags(tags, maxItems = 20) {
    if (!Array.isArray(tags)) {
        return { valid: false, error: 'Tags must be an array' };
    }
    if (tags.length > maxItems) {
        return { valid: false, error: `Cannot use more than ${maxItems} tags` };
    }
    
    for (const tag of tags) {
        if (typeof tag !== 'string') {
            return { valid: false, error: 'Each tag must be a string' };
        }
        if (tag.length === 0 || tag.length > 50) {
            return { valid: false, error: 'Each tag must be between 1 and 50 characters' };
        }
        if (!/^[a-zA-Z0-9_#-]+$/.test(tag)) {
            return { valid: false, error: 'Tags can only contain letters, numbers, underscores, hyphens, and #' };
        }
    }
    
    return { valid: true };
}

/**
 * Validate visibility setting
 * @param {string} visibility - Visibility setting
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateVisibility(visibility) {
    const validValues = ['public', 'followers', 'private'];
    if (!visibility) {
        return { valid: false, error: 'Visibility is required' };
    }
    if (!validValues.includes(visibility)) {
        return { valid: false, error: `Visibility must be one of: ${validValues.join(', ')}` };
    }
    return { valid: true };
}

/**
 * Validate message content (for Thesidia API)
 * @param {string} message - Message content
 * @param {number} maxLength - Maximum length (default 10000)
 * @returns {Object} {valid: boolean, error?: string}
 */
export function validateMessage(message, maxLength = 10000) {
    if (!message) {
        return { valid: false, error: 'Message is required' };
    }
    if (typeof message !== 'string') {
        return { valid: false, error: 'Message must be a string' };
    }
    if (message.length > maxLength) {
        return { valid: false, error: `Message must be no more than ${maxLength} characters` };
    }
    if (message.trim().length === 0) {
        return { valid: false, error: 'Message cannot be empty' };
    }
    return { valid: true };
}

